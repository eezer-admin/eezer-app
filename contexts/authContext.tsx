import { createContext, useEffect, useState } from 'react';

import { User } from '@src/domain/entities/User';
import { ClearUserUseCase } from '@usecases/user/ClearUserUseCase';
import { GetUserUseCase } from '@usecases/user/GetUserUseCase';
import { LoginUseCase } from '@usecases/user/LoginUseCase';
import { removeFromStorage as removeTransportLog } from '../services/TransportLogService';
import { remove as removeCurrentTransport } from '../services/TransportService';
import { AuthContextData } from '../types/Auth';

export const AuthContext = createContext<AuthContextData>({} as AuthContextData);

export const AuthProvider = (props) => {
  const [user, setUser] = useState<AuthContextData>();
  const [authLoaded, setAuthLoaded] = useState(false);

  // Determine if the user is already logged in when the app loads.
  // If there's a user in the local storage, assign it to the context
  // to make the app recognize it.
  useEffect(() => {
    new GetUserUseCase().execute().then((user: User | null) => {
      if (user) {
        setUser(user);
      }
      setAuthLoaded(true);
    });
  }, []);

  // Makes an actual login request towards the backend API.
  const login = async (username: string, password: string): Promise<User | null> => {
    return new LoginUseCase().execute(username, password).then((user: User) => {
      if (user) {
        setUser(user);

        return user;
      }

      return null;
    });
  };

  // Removes the user from the local storage and context.
  const logout = async () => {
    await removeTransportLog();
    await removeCurrentTransport();
    await new ClearUserUseCase().execute();

    setUser(null);
  };

  const isLoggedIn = (): boolean => {
    return !!user?.access_token;
  };

  return (
    <AuthContext.Provider value={{ user, authLoaded, login, logout, isLoggedIn }}>
      {props.children}
    </AuthContext.Provider>
  );
};
