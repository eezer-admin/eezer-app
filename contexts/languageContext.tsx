import { createContext, useEffect, useState } from 'react';

import i18n, { defaultLanguage } from '../localization/Localization';

export const LanguageContext = createContext<string>(defaultLanguage());

export const LanguageProvider = (props) => {
  const [language, setLanguage] = useState(props.language);

  // Update the i18n locale config before re-rendering the app.
  useEffect(() => {
    i18n.locale = language;
  }, [language]);

  return (
    <LanguageContext.Provider value={[language, setLanguage]}>
      {props.children}
    </LanguageContext.Provider>
  );
};
