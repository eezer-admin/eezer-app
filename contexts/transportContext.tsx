import { createContext, useEffect, useState } from 'react';

import TransportModel from '../models/TransportModel';
import { get as getFromStorage, persist, remove } from '../services/TransportService';
import { TransportContextData } from '../types/Transports';

export const TransportContext = createContext<TransportContextData>({} as TransportContextData);

export const TransportProvider = (props) => {
  const [data, setData] = useState<TransportContextData>();

  useEffect(() => {
    getFromStorage().then((transport: TransportModel) => {
      if (transport) {
        setData(transport);
      }
    });
  }, []);

  const get = async (): Promise<TransportModel | null> => {
    return getFromStorage();
  };

  const start = async (reason: string, vehicle_id: number): Promise<TransportModel> => {
    const transport = new TransportModel(null);
    transport.start(reason);
    transport.setVehicleId(vehicle_id);
    setData(transport);

    return persist(transport);
  };

  const completeTransport = async (): Promise<void> => {
    setData(null);

    return remove();
  };

  const save = async (transport: TransportModel): Promise<TransportModel> => {
    return persist(transport).then(() => {
      setData(transport);

      return transport;
    });
  };

  return (
    <TransportContext.Provider
      value={{
        data,
        get,
        save,
        start,
        completeTransport,
      }}>
      {props.children}
    </TransportContext.Provider>
  );
};
