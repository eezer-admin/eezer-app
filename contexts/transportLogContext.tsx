import { createContext, useEffect, useState } from 'react';

import { add, get as getFromStorage } from '../services/TransportLogService';
import { TransportLog, TransportLogContextData } from '../types/Transports';

export const TransportLogContext = createContext<TransportLogContextData>({} as TransportLogContextData);

export const TransportLogProvider = (props) => {
  const [data, setData] = useState<TransportLogContextData>();

  useEffect(() => {
    getFromStorage().then((log: TransportLog) => {
      if (log) {
        setData(log);
      }
    });
  }, []);

  return <TransportLogContext.Provider value={[data, setData]}>{props.children}</TransportLogContext.Provider>;
};
