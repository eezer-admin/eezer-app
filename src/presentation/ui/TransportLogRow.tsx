import { Transport } from '@src/domain/entities/Transport';
import { useContext, useState } from 'react';
import { Text, TouchableOpacity, View } from 'react-native';

import { AuthContext } from '../../../contexts/authContext';
import { __ } from '../../../localization/Localization';
import Styles from '../../../styles/Styles';

export type TransportLogRowProps = {
  transport: Transport;
};

export default function TransportLogRow(props: TransportLogRowProps) {
  const auth = useContext(AuthContext);
  const [showDetails, setShowDetails] = useState(false);

  return (
    <View
      style={{
        borderBottomColor: Styles.colors.white,
        borderBottomWidth: 5,
      }}>
      <TouchableOpacity
        onPress={() => {
          setShowDetails(!showDetails);
        }}
        style={{
          backgroundColor: showDetails ? Styles.colors.grayDark : null,
          paddingVertical: Styles.margins.large,
          paddingHorizontal: Styles.container.paddingHorizontal,
          flex: 1,
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
        }}>
        <Text style={{ width: '33%' }}>{props.transport.getReadableStartDate()}</Text>
        <Text style={{ width: '33%', textAlign: 'center' }}>{props.transport.getReadableDuration()}</Text>
        <Text style={{ width: '33%', textAlign: 'right' }}>{props.transport.isSynced() ? __('Yes') : __('No')}</Text>
      </TouchableOpacity>

      {showDetails ? (
        <View
          style={{
            paddingVertical: Styles.margins.large,
            paddingHorizontal: Styles.container.paddingHorizontal,
            backgroundColor: Styles.colors.gray,
            flex: 1,
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
          }}>
          <View style={{ width: '33%' }}>
            <Text style={{ fontSize: Styles.fontSizes.small }}>{__('Driver')}</Text>
            <Text style={{ fontSize: Styles.fontSizes.small }}>
              {auth.user?.first_name} {auth.user?.last_name}
            </Text>
          </View>
          <View style={{ width: '33%' }}>
            <Text style={{ textAlign: 'center', fontSize: Styles.fontSizes.small }}>{__('Distance')}</Text>
            <Text style={{ textAlign: 'center', fontSize: Styles.fontSizes.small }}>
              {props.transport.getReadableDistance()}
            </Text>
          </View>
          <View style={{ width: '33%' }}>
            <Text style={{ textAlign: 'right', fontSize: Styles.fontSizes.small }}>{__('Cause')}</Text>
            <Text style={{ textAlign: 'right', fontSize: Styles.fontSizes.small }}>{props.transport.reason}</Text>
          </View>
        </View>
      ) : null}
    </View>
  );
}
