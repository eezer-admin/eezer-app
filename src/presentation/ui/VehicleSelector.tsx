import * as React from 'react';
import { useContext } from 'react';
import { Text, TouchableOpacity, View } from 'react-native';

import { AuthContext } from '../../../contexts/authContext';
import { __ } from '../../../localization/Localization';
import Styles from '../../../styles/Styles';

type VehicleSelectorProps = {
  onNoVehiclesAvailable(): void;
  onVehicleSelected(vehicleId: number): void;
  vehicleId: number | null;
};

export default function VehicleSelector(props: VehicleSelectorProps) {
  const auth = useContext(AuthContext);

  if (!auth.user?.vehicles?.length) {
    props.onNoVehiclesAvailable();

    return (
      <View
        style={{
          padding: Styles.margins.large,
          margin: Styles.margins.medium,
          backgroundColor: Styles.colors.redLight,
          borderRadius: 8,
        }}>
        <Text>{__('No vehicles available. Contact an administrator.')}</Text>
      </View>
    );
  }

  return (
    <View>
      <Text style={{ fontSize: Styles.fontSizes.default }}>{__('Select vehicle')}</Text>
      <View style={{ flexDirection: 'row' }}>
        {auth.user.vehicles.map((vehicle) => (
          <TouchableOpacity
            style={{
              padding: Styles.margins.large,
              margin: Styles.margins.medium,
              width: 100,
              height: 80,
              borderColor: Styles.colors.gray,
              borderWidth: 2,
              justifyContent: 'center',
              alignItems: 'center',
              backgroundColor: props.vehicleId === vehicle.id ? Styles.colors.gray : 'transparent',
            }}
            key={`vehicle-${vehicle.id}`}
            onPress={() => {
              props.onVehicleSelected(vehicle.id);
            }}>
            <Text style={{ fontSize: Styles.fontSizes.medium }}>{vehicle.name}</Text>
          </TouchableOpacity>
        ))}
      </View>
    </View>
  );
}
