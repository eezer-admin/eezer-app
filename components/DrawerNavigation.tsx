import { DrawerContentScrollView, DrawerItemList } from '@react-navigation/drawer';
import * as React from 'react';
import { useContext } from 'react';
import { Image, Text, TouchableOpacity, View } from 'react-native';

import { AuthContext } from '../contexts/authContext';
import { LanguageContext } from '../contexts/languageContext';
import { __ } from '../localization/Localization';
import Styles from '../styles/Styles';

export const DrawerNavigation = (props) => {
  const auth = useContext(AuthContext);
  const [language, setLanguage] = useContext(LanguageContext);

  return (
    <View style={{ flex: 1 }}>
      <DrawerContentScrollView {...props}>
        <DrawerItemList {...props} />
      </DrawerContentScrollView>

      <View
        style={{
          borderTopColor: Styles.colors.gray,
          borderTopWidth: 1,
          paddingHorizontal: Styles.margins.medium,
          paddingVertical: 30,
        }}>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'sw' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            setLanguage('sw');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image style={{ width: 20, height: 13 }} source={require('../assets/flag-tz.png')} />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'sw' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('Kiswahili')}
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'en' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            setLanguage('en');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image style={{ width: 20, height: 13 }} source={require('../assets/flag-en.png')} />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'en' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('English')}
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'fr' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            setLanguage('fr');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image style={{ width: 20, height: 13 }} source={require('../assets/flag-fr.png')} />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'fr' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('French')}
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'rn' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            setLanguage('rn');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image style={{ width: 20, height: 13 }} source={require('../assets/flag-rn.png')} />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'rn' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('Kirundi')}
            </Text>
          </View>
        </TouchableOpacity>
      </View>

      {auth.isLoggedIn() ? (
        <View
          style={{
            borderTopColor: Styles.colors.gray,
            borderTopWidth: 1,
            paddingHorizontal: 20,
            paddingVertical: 30,
          }}>
          <TouchableOpacity
            style={{
              paddingVertical: Styles.margins.medium,
            }}
            onPress={() => {
              auth.logout().then(() => {
                props.navigation.closeDrawer();
              });
            }}>
            <Text style={{ ...Styles.text.default }}>{__('Logout')}</Text>
          </TouchableOpacity>
        </View>
      ) : null}
    </View>
  );
};
