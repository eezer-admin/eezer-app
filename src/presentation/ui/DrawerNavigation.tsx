import { DrawerContentScrollView, DrawerItemList } from '@react-navigation/drawer';
import * as React from 'react';
import { useContext } from 'react';
import { Text, TouchableOpacity, View } from 'react-native';

import Constants from 'expo-constants';
import { AuthContext } from '../../../contexts/authContext';
import { LanguageContext } from '../../../contexts/languageContext';
import { __ } from '../../../localization/Localization';
import Styles from '../../../styles/Styles';

export const DrawerNavigation = (props) => {
  const auth = useContext(AuthContext);
  const [language, setLanguage] = useContext(LanguageContext);

  return (
    <View style={{ flex: 1 }}>
      <DrawerContentScrollView {...props}>
        <DrawerItemList {...props} />
      </DrawerContentScrollView>

      <Text
        style={{
          fontSize: Styles.fontSizes.small,
          color: Styles.colors.grayDarker,
          margin: Styles.margins.medium,
          textAlign: 'right',
        }}>
        v{Constants.expoConfig?.version}-{Constants.expoConfig?.android?.versionCode || null}
      </Text>
      {/* Temporarily disabled because of bug causing crash when changing language the pressing Android back button.
      <View
        style={{
          borderTopColor: Styles.colors.gray,
          borderTopWidth: 1,
          paddingHorizontal: Styles.margins.medium,
          paddingVertical: 30,
        }}>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'sw' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            updateLocale('sw');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image
              style={{ width: 20, height: 13 }}
              source={require('../../../assets/flag-tz.png')}
            />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'sw' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('Kiswahili')}
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'en' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            updateLocale('en');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image
              style={{ width: 20, height: 13 }}
              source={require('../../../assets/flag-en.png')}
            />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'en' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('English')}
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'fr' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            updateLocale('fr');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image
              style={{ width: 20, height: 13 }}
              source={require('../../../assets/flag-fr.png')}
            />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'fr' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('French')}
            </Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity
          style={{
            paddingVertical: Styles.margins.medium,
            padding: Styles.margins.medium,
            backgroundColor: language === 'rn' ? '#E0EFFF' : null,
            borderRadius: 5,
          }}
          onPress={() => {
            updateLocale('rn');
          }}>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image
              style={{ width: 20, height: 13 }}
              source={require('../../../assets/flag-rn.png')}
            />
            <Text
              style={{
                marginLeft: Styles.margins.medium,
                color: language === 'rn' ? '#007AFF' : Styles.colors.grayDarker,
              }}>
              {__('Kirundi')}
            </Text>
          </View>
        </TouchableOpacity>
      </View>
       */}

      {auth.isLoggedIn() ? (
        <View
          style={{
            borderTopColor: Styles.colors.gray,
            borderTopWidth: 1,
            paddingHorizontal: 20,
            paddingVertical: 30,
          }}>
          <TouchableOpacity
            style={{
              paddingVertical: Styles.margins.medium,
            }}
            onPress={() => {
              auth.logout().then(() => {
                props.navigation.closeDrawer();
              });
            }}>
            <Text style={{ ...Styles.text.default }}>{__('Logout')}</Text>
          </TouchableOpacity>
        </View>
      ) : null}
    </View>
  );
};
