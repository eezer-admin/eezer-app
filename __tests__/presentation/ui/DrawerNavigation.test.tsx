import { describe, expect, it } from '@jest/globals';
import { DrawerNavigation } from '@presentation/ui/DrawerNavigation';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import renderer from 'react-test-renderer';

import { AuthProvider } from '../../../contexts/authContext';

describe('DrawerNavigation', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <SafeAreaProvider>
          <AuthProvider>
            <DrawerNavigation />
          </AuthProvider>
        </SafeAreaProvider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
