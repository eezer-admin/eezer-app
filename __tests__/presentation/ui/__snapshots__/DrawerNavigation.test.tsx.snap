// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerNavigation renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
